casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -411,4 +418,48 @@ object ProtoUtil {
     (missingParents union missingJustifications).toList
   }
 
+  // Return hashes of all blocks that are yet to be seen by the passed in block
+  def unseenBlockHashes[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMessage
+  ): F[Set[BlockHash]] =
+    for {
+      latestMessages        <- dag.latestMessages
+      latestMessagesOfBlock <- ProtoUtil.toLatestMessage(block.justifications, dag)
+      result <- latestMessages.toList
+                 .traverse {
+                   case (validator, latestMessage) =>
+                     val lastSeenBlockHash = latestMessagesOfBlock(validator)
+                     DagOperations
+                       .bfTraverseF(List(latestMessage))(
+                         block => getCreatorJustificationUnlessGoal(dag, block, lastSeenBlockHash)
+                       )
+                       .map(_.blockHash)
+                       .toSet
+                 }
+                 .map(_.flatten.toSet)
+    } yield result -- latestMessagesOfBlock.values.map(_.blockHash) - block.blockHash
+
+  private def getCreatorJustificationUnlessGoal[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMetadata,
+      goal: BlockMetadata
+  ): F[List[BlockMetadata]] =
+    ProtoUtil.creatorJustification(block) match {
+      case Some(Justification(_, hash)) =>
+        dag.lookup(hash).flatMap {
+          case Some(creatorJustification) =>
+            if (creatorJustification == goal) {
+              List.empty[BlockMetadata].pure[F]
+            } else {
+              List(creatorJustification).pure[F]
+            }
+          case None =>
+            Sync[F].raiseError[List[BlockMetadata]](
+              new RuntimeException(s"Missing block hash $hash in block dag.")
+            )
+        }
+      case None =>
+        List.empty[BlockMetadata].pure[F]
+    }
```

Comment:
This function with two layers of matches is unreadable. The version below reads better and should work the same

```scala
  private def getCreatorJustificationUnlessGoal[F[_]: Sync: BlockStore](
      dag: BlockDagRepresentation[F],
      block: BlockMetadata,
      goal: BlockMetadata
  ): F[List[BlockMetadata]] = {
    def findByHash(hash : ByteString) =
      for {
        maybeBlock  <- dag.lookup(hash)
        maybeFoundBlock = maybeBlock.filter(_ == goal)
        result <- maybeFoundBlock
          .fold(
            Sync[F].raiseError[Option[BlockMetadata]](new RuntimeException(s"Missing block hash $hash in block dag.")))(
            j => Option(j).pure[F])
      } yield result

    ProtoUtil.creatorJustification(block)
      .map(_.latestBlockHash)
      .traverse(findByHash)
      .map(_.flatten.toList)
  }
```

Reply:
Personally, I find your version much harder to read. In fact it isn't the same - when the creatorJustification is None, it should return an empty list.

Reply:
OK. If the behavior of my code is different then the case is closed :)